# Socrates2



`Socrates2` is an R and python package for analyzing Single-Cell Assay for Transposase Accessible Chromatin Sequencing (scATAC-seq) data. `Socrates2` takes the BAM file built from cellranger as the inputs to guild users to conduct a comprehensive scATAC-seq analysis. It updated several functions compared to the `Socrates` to help users perform an end-to-end analysis. 



## Installation
For the installation manual, please refer to the scPlantReg platform: http://112.126.28.109:8150/pipeline.



## Tutorials
Before running Socrates2, users need to prepare aligned BAM files as input for all downstream analyses.
We provide detailed tutorials for generating BAM files from scATAC-seq data produced by multiple experimental platforms:

[1. 10X Genomics Chromium platform](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/links_to_different_data_inputs/10x-Genomics_scATAC-seq_Data_Processing_Bam.html)

[2. MGI DNBseq platform](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/links_to_different_data_inputs/BGI-DNBseq_scATAC-seq_Data_Processing_Bam.html)

[3. BioRad platform](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/links_to_different_data_inputs/BioRad_scATAC-seq_Data_Processing_Bam.html)

[4. Singleron platform](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/links_to_different_data_inputs/Singleron_scATAC-seq_Data_Processing_Bam.html)

After obtaining the BAM files, please follow the step-by-step tutorials below to perform a complete end-to-end Socrates2 analysis workflow:

[1. Processing bam](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/process_BAM.html)

[2. Load and QC](https://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/load_data_and_QC.html)

[3. Clustering](https://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/clustering.html)

[4. Prepare gene accessibility](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/prepare_gene_accessibility.html)

[5. Find diff markers](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/find_diff_markers.html)

[6. Annotate cell identity](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/annotate_cell_identity.html)

[7. Call peak per cell type](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/peak_call_per_celltype.html)

[8. Call diff peaks](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/call_diff_peaks.html)

[9. Motif analysis](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/motif_analysis.html)

[10. Trajectory analysis](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/trajectory_analysis.html)

[11. Predict activating&repressing ACR](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/predict_activating_repressing_ACR.html)

[12. Cross species analysis](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/cross_species_analysis.html)


## Issue reports
For a summary of the most common issues and their corresponding solutions, please refer to the following documentation: [Issue Reports](http://htmlpreview.github.io/?https://github.com/yanhaidong1/Socrates2/blob/main/vignettes/issue_reports.html)









...  more coming soon
